{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d8d92ed1-fa98-4d56-bc3e-4cc8f20870af",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "import yaml\n",
    "import numpy as np\n",
    "\n",
    "# load in config with GPSC paths\n",
    "with open(\"config/SP_prokka_roary.yaml\", 'r') as file:\n",
    "    config=yaml.safe_load(file)\n",
    "\n",
    "# extract GPSCs\n",
    "gpscs = config['samples'] \n",
    "\n",
    "# set params\n",
    "amplicon_stats = list()\n",
    "xlen = 2200\n",
    "\n",
    "total_genome_coverages={}\n",
    "amplicon_positions={}\n",
    "\n",
    "with open('output.txt', 'w') as f:\n",
    "\n",
    "    for gpsc, fasta_file in gpscs.items():\n",
    "        records = list(SeqIO.parse(fasta_file, \"fasta\"))\n",
    "\n",
    "        # calculate length of the genome for each GPSC\n",
    "        genome_length=sum(len(record.seq) for record in records)\n",
    "\n",
    "        # initialize amplicon genome coverage for each GPSC separately \n",
    "        total_genome_coverage=0\n",
    "\n",
    "        # initialize set of covered positions across each GPSC \n",
    "        covered_positions = set()\n",
    "\n",
    "        # intitialize list of amplicon positions \n",
    "        amplicon_positions[gpsc] = []\n",
    "\n",
    "        print(f\"Processing {gpsc}...\", file=f)  # Print the name of the GPSC being processed\n",
    "\n",
    "        # Load the samtools depth file\n",
    "        depth_file = os.path.join(\"depth\", f\"{gpsc}.depth\")\n",
    "        df = pd.read_csv(depth_file, sep=\"\\t\", names=[\"Ref\", \"Pos\", \"Depth\"])\n",
    "\n",
    "        # Get the positions where the depth is 1 (primer binding sites)\n",
    "        primer_binding_sites = df[df[\"Depth\"] == 1][\"Pos\"].tolist()\n",
    "\n",
    "        for p1loc in primer_binding_sites:\n",
    "            # Find the next primer binding site within xlen bases\n",
    "            p2loc = next((pos for pos in primer_binding_sites if p1loc < pos <= p1loc + xlen), None)\n",
    "\n",
    "            if p2loc is not None:\n",
    "                # Calculate the amplicon stats\n",
    "                amplicon_stats.append((gpsc, gpsc, gpsc, gpsc, gpsc, gpsc, 0, 0, 0, p1loc, p2loc))\n",
    "                covered_positions.update(range(p1loc, p2loc+1))\n",
    "                # store amplicon positions \n",
    "                amplicon_positions[gpsc].append((p1loc, p2loc))\n",
    "                print(f\"Detected amplicon from {p1loc} to {p2loc}.\", file=f)  # Print the start and end positions of the detected amplicon\n",
    "\n",
    "                # find total length of the amplicon\n",
    "                total_genome_coverage += p2loc-p1loc # calculate the length of the amplicon and add to total genome coverage\n",
    "\n",
    "        # calculate predicted % coverage \n",
    "        coverage_percentage = (len(covered_positions) / genome_length) * 100\n",
    "\n",
    "        # update dictionary \n",
    "        total_genome_coverages[gpsc] = coverage_percentage\n",
    "\n",
    "        print(f\"Total genome coverage for {gpsc}: {total_genome_coverages[gpsc]}%\", file=f)  # Print the total genome coverage for the GPSC\n",
    "\n",
    "colnames=[\"pid1\",\"pid2\",\"set1\",\"set2\",\n",
    "          \"pseq1\",\"pseq2\",\n",
    "          \"max_hdist\",\"hdist1\",\"hdist2\",\n",
    "          \"p1loc\",\"p2loc\"]\n",
    "coltypes=[\"<U30\", \"<U30\", \"<U30\", \"<U30\",\n",
    "          \"<U30\", \"<U30\",\n",
    "          float, float, float,\n",
    "          int, int]\n",
    "\n",
    "dt = {'names':colnames, 'formats':coltypes}\n",
    "\n",
    "amplicon_statsnp = np.array(amplicon_stats,\n",
    "                     dtype=dt)\n",
    "\n",
    "np.save(\"amplicon_statstab.npy\", amplicon_statsnp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9ce5ee31-fe86-4cc1-aa67-4b2c9386bfd7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GPSC12: 99.28831877584545%\n",
      "GPSC25: 83.70355584570514%\n",
      "GPSC31: 83.420166858458%\n",
      "GPSC34: 86.58638733761124%\n",
      "GPSC3: 82.58450750499864%\n",
      "GPSC4: 82.57514313707841%\n",
      "JYGP01: 37.48834324179536%\n",
      "GPSC17: 85.52519860790558%\n",
      "GPSC27: 81.42039075508528%\n",
      "GPSC32: 86.56638367738053%\n",
      "GPSC37: 82.37399570281472%\n",
      "GPSC41: 86.57502063559636%\n",
      "GPSC5: 82.9434782199731%\n",
      "NC_017592: 99.28831877584545%\n"
     ]
    }
   ],
   "source": [
    "for gpsc, coverage in total_genome_coverages.items():\n",
    "    print(f\"{gpsc}: {coverage}%\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1e9aac56-1f9b-42e6-b335-98bd07c7bc47",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GPSC12: 2036867 bp\n",
      "GPSC25: 2129170 bp\n",
      "GPSC31: 2085600 bp\n",
      "GPSC34: 2052251 bp\n",
      "GPSC3: 2130580 bp\n",
      "GPSC4: 2121917 bp\n",
      "JYGP01: 1915198 bp\n",
      "GPSC17: 2107041 bp\n",
      "GPSC27: 2160842 bp\n",
      "GPSC32: 2053788 bp\n",
      "GPSC37: 2133955 bp\n",
      "GPSC41: 2001396 bp\n",
      "GPSC5: 2126260 bp\n",
      "NC_017592: 2036867 bp\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "# Load the config file\n",
    "with open(\"config/SP_prokka_roary.yaml\", 'r') as file:\n",
    "    config=yaml.safe_load(file)\n",
    "\n",
    "# Extract the GPSCs\n",
    "gpscs = config['samples']\n",
    "\n",
    "# Initialize a dictionary to store the genome lengths\n",
    "genome_lengths = {}\n",
    "\n",
    "# Calculate and store the length of each sequence\n",
    "for gpsc, fasta_file in gpscs.items():\n",
    "    records = list(SeqIO.parse(fasta_file, \"fasta\"))\n",
    "    genome_length = sum(len(record.seq) for record in records)\n",
    "    genome_lengths[gpsc] = genome_length  # Store the genome length in the dictionary\n",
    "\n",
    "#Print the genome lengths\n",
    "for gpsc, length in genome_lengths.items():\n",
    "    print(f\"{gpsc}: {length} bp\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6a27d6f-d02c-420d-9dd3-0d5c1b3229d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Determine the number of rows and columns for the subplots\n",
    "n = len(amplicon_positions)\n",
    "ncols = 1\n",
    "nrows = n // ncols + (n % ncols > 0)\n",
    "\n",
    "# Create a figure and a grid of subplots\n",
    "fig, axs = plt.subplots(nrows, ncols, figsize=(10, 5*nrows))\n",
    "\n",
    "# Flatten the array of axes, in case we have less subplots than planned\n",
    "axs = axs.flatten()\n",
    "\n",
    "# Plot the data for each fasta sequence in a separate subplot\n",
    "for i, (fasta_name, positions) in enumerate(amplicon_positions.items()):\n",
    "    for start, end in positions:\n",
    "        axs[i].fill_between([start, end], [1, 1], color='purple')\n",
    "    axs[i].set_title(\"Coverage for fasta sequence {}\".format(fasta_name))\n",
    "    axs[i].set_xlabel(\"Position\")\n",
    "    axs[i].set_xlim(0, genome_lengths[fasta_name])  # Set x limit to the length of the genome\n",
    "    axs[i].set_ylim(0, 1)\n",
    "\n",
    "# Remove empty subplots\n",
    "for i in range(n, nrows*ncols):\n",
    "    fig.delaxes(axs[i])\n",
    "\n",
    "# Adjust the layout\n",
    "plt.tight_layout()\n",
    "\n",
    "# Save the figure as a PNG file\n",
    "plt.savefig(\"fasta_coverage.png\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "646b08cb-6b15-4447-af10-b34014e0eaa6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "\n",
    "with open('amplicon_positions.pkl', 'wb') as f:\n",
    "    pickle.dump(amplicon_positions, f)\n",
    "\n",
    "with open('genome_coverage_pc.pkl', 'wb') as f:\n",
    "    pickle.dump(total_genome_coverages, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f1fa2583-ae16-485e-9767-8208b013d8b6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.065875589\n"
     ]
    }
   ],
   "source": [
    "# get the branch length of the outgroup since it isn't plotting right when i root it\n",
    "\n",
    "from Bio import Phylo\n",
    "\n",
    "# Read the tree from the Newick file\n",
    "tree = Phylo.read('GPSC_SP_reps.newick', 'newick')\n",
    "\n",
    "# Get the clade corresponding to the outgroup\n",
    "outgroup = tree.common_ancestor('JYGP01')\n",
    "\n",
    "# Get the original branch length of the outgroup\n",
    "original_branch_length = outgroup.branch_length\n",
    "\n",
    "print(original_branch_length)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac9a8aa9-535d-40a5-a1de-4f7156349401",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib\n",
    "matplotlib.use('TkAgg') \n",
    "\n",
    "from Bio import Phylo\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Read the tree from the Newick file\n",
    "tree = Phylo.read('GPSC_SP_reps.newick', 'newick')\n",
    "\n",
    "# Reroot the tree at the outgroup\n",
    "tree.root_with_outgroup('JYGP01')\n",
    "\n",
    "# Get the clade corresponding to the outgroup\n",
    "outgroup = tree.common_ancestor('JYGP01')\n",
    "\n",
    "# Manually set the branch length of the outgroup\n",
    "outgroup.branch_length = 0.063920664\n",
    "\n",
    "# Draw the tree\n",
    "Phylo.draw(tree, axes=ax)\n",
    "\n",
    "# Show the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "1f160bf1-f725-4f89-becd-943a8202343a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tree(rooted=True, weight=1.0)\n",
      "    Clade()\n",
      "        Clade(branch_length=0.065875589, confidence=1.0)\n",
      "            Clade(branch_length=0.001565992, confidence=1.0)\n",
      "                Clade(branch_length=0.001067135, confidence=1.0)\n",
      "                    Clade(branch_length=0.000500928)\n",
      "                        Clade(branch_length=0.001405424, confidence=1.0)\n",
      "                            Clade(branch_length=0.004816411, name='GPSC17')\n",
      "                            Clade(branch_length=0.004338649, name='GPSC37')\n",
      "                        Clade(branch_length=0.000397671, confidence=0.922)\n",
      "                            Clade(branch_length=0.001072732, confidence=1.0)\n",
      "                                Clade(branch_length=0.004405656)\n",
      "                                    Clade(branch_length=0.0, name='GPSC12')\n",
      "                                    Clade(branch_length=0.0, name='NC_017592')\n",
      "                                Clade(branch_length=0.004314957, name='GPSC3')\n",
      "                            Clade(branch_length=0.000748469, confidence=1.0)\n",
      "                                Clade(branch_length=0.004716108, name='GPSC34')\n",
      "                                Clade(branch_length=0.001170461, confidence=1.0)\n",
      "                                    Clade(branch_length=0.004536119, name='GPSC25')\n",
      "                                    Clade(branch_length=0.004145008, name='GPSC41')\n",
      "                    Clade(branch_length=0.000491534, confidence=1.0)\n",
      "                        Clade(branch_length=0.004989177, name='GPSC32')\n",
      "                        Clade(branch_length=0.001100372, confidence=0.969)\n",
      "                            Clade(branch_length=0.004250587, name='GPSC27')\n",
      "                            Clade(branch_length=0.004319177, name='GPSC4')\n",
      "                Clade(branch_length=0.004845407, name='GPSC31')\n",
      "            Clade(branch_length=0.003560917, name='GPSC5')\n",
      "        Clade(branch_length=0.065875589, name='JYGP01')\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/v8/xq20p7kj2kd8tzvx4xtdhv65d3pw82/T/ipykernel_5746/3980985191.py:85: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`\n",
      "  roary.replace('.{2,100}', 1, regex=True, inplace=True)\n"
     ]
    }
   ],
   "source": [
    "# Plot the predicted amplicon coverage with the percent coverage to the right \n",
    "\n",
    "def get_options():\n",
    "    import argparse\n",
    "\n",
    "    # Define the command line arguments as variables\n",
    "    tree = 'GPSC_SP_reps.newick'\n",
    "    spreadsheet = 'gene_presence_absence.csv'\n",
    "    depth_dir = '/vast/palmer/scratch/turner/flg9/snakemake_workflows/pangenome_alignment/data/results/samtools_depth'\n",
    "    labels = False\n",
    "    format = 'png'\n",
    "    skipped_columns = 14\n",
    "\n",
    "    # create the top-level parser\n",
    "    description = \"Create plots from roary outputs\"\n",
    "    parser = argparse.ArgumentParser(description = description,\n",
    "                                     prog = 'roary_plots.py')\n",
    "\n",
    "    parser.add_argument('tree', action='store',\n",
    "                        help='Newick Tree file', default=tree)\n",
    "    parser.add_argument('spreadsheet', action='store',\n",
    "                        help='Roary gene presence/absence spreadsheet', default=spreadsheet)\n",
    "    parser.add_argument('--labels', action='store_true',\n",
    "                        default=labels,\n",
    "                        help='Add node labels to the tree (up to 10 chars)')\n",
    "    parser.add_argument('--format',\n",
    "                        choices=('png',\n",
    "                                 'tiff',\n",
    "                                 'pdf',\n",
    "                                 'svg'),\n",
    "                        default=format,\n",
    "                        help='Output format [Default: png]')\n",
    "    parser.add_argument('-N', '--skipped-columns', action='store',\n",
    "                        type=int,\n",
    "                        default=skipped_columns,\n",
    "                        help='First N columns of Roary\\'s output to exclude [Default: 14]')\n",
    "\n",
    "    return argparse.Namespace(tree=tree, spreadsheet=spreadsheet, depth_dir=depth_dir, labels=labels, format=format, skipped_columns=skipped_columns)\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    options = get_options()\n",
    "\n",
    "    import matplotlib\n",
    "    from matplotlib.patches import Rectangle\n",
    "    from matplotlib.colors import LinearSegmentedColormap\n",
    "    matplotlib.use('Agg')\n",
    "\n",
    "    import matplotlib.pyplot as plt\n",
    "    import seaborn as sns\n",
    "\n",
    "    sns.set_style('white')\n",
    "\n",
    "    import os\n",
    "    import pandas as pd\n",
    "    import numpy as np\n",
    "    from Bio import Phylo\n",
    "    import pickle \n",
    "\n",
    "    # Read the tree from the Newick file\n",
    "    t = Phylo.read(options.tree, 'newick')\n",
    "\n",
    "    # Reroot the tree at the outgroup\n",
    "    t.root_with_outgroup('JYGP01')\n",
    "\n",
    "    # Get the clade corresponding to the outgroup\n",
    "    outgroup = t.common_ancestor('JYGP01')\n",
    "\n",
    "    # Manually set the branch length of the outgroup\n",
    "    outgroup.branch_length = 0.065875589  # replace this with the original branch length\n",
    "\n",
    "    # check that re-rooting and manual branching occurred \n",
    "    print(t)\n",
    "\n",
    "    # Max distance to create better plots\n",
    "    mdist = max([t.distance(t.root, x) for x in t.get_terminals()])\n",
    "\n",
    "    # Load roary\n",
    "    roary = pd.read_csv(options.spreadsheet, low_memory=False)\n",
    "    # Set index (group name)\n",
    "    roary.set_index('Gene', inplace=True)\n",
    "    # Drop the other info columns\n",
    "    roary.drop(list(roary.columns[:options.skipped_columns-1]), axis=1, inplace=True)\n",
    "\n",
    "    # Transform it in a presence/absence matrix (1/0)\n",
    "    roary.replace('.{2,100}', 1, regex=True, inplace=True)\n",
    "    roary.replace(np.nan, 0, regex=True, inplace=True)\n",
    "\n",
    "    # Sort the matrix by the sum of strains presence\n",
    "    idx = roary.sum(axis=1).sort_values(ascending=False).index\n",
    "    roary_sorted = roary.loc[idx]\n",
    "\n",
    "    # Sort the matrix according to tip labels in the tree\n",
    "    roary_sorted = roary_sorted[[x.name for x in t.get_terminals()]]\n",
    "\n",
    "    # Calculate the number of terminals\n",
    "    num_terminals = len(t.get_terminals())\n",
    "\n",
    "    # Adjust the number of columns in the grid to accommodate the number of terminals\n",
    "    num_columns = max(60, 3 * num_terminals)\n",
    "    \n",
    "    # Plot presence/absence matrix against the tree\n",
    "    # Create a custom colormap\n",
    "    cmap = LinearSegmentedColormap.from_list(\"my_colormap\", [\"white\", \"#93a3b1\"])\n",
    "\n",
    "    with sns.axes_style('whitegrid'):\n",
    "        fig = plt.figure(figsize=(35, 20))  # Adjusted figure size\n",
    "\n",
    "        ax1 = plt.subplot2grid((1, num_columns), (0, num_columns//3), colspan=num_columns//3)  # Matrix takes up 1/3 of the figure\n",
    "        a = ax1.matshow(roary_sorted.T, cmap=cmap,\n",
    "                   vmin=0, vmax=1,\n",
    "                   aspect='auto',\n",
    "                   interpolation='none'\n",
    "                    )\n",
    "        ax1.set_yticks([])\n",
    "        ax1.set_xticks([])\n",
    "        ax1.axis('off')\n",
    "        ax1.set_title('Roary matrix\\n(%d gene clusters)'%roary.shape[0], fontsize = 12)\n",
    "\n",
    "        # Calculate the number of rows each subplot should span\n",
    "        num_rows = len(t.get_terminals())\n",
    "        rows_per_subplot = max(1, num_rows // len(t.get_terminals()))\n",
    "\n",
    "        # Create subplots for coverage depth plots\n",
    "        axs_depth = [plt.subplot2grid((num_rows, num_columns), (i * rows_per_subplot, 2*num_columns//3), rowspan=rows_per_subplot, colspan=num_columns//3) for i in range(num_terminals)]  # Depth plots take up 1/3 of the figure\n",
    "\n",
    "        ax = plt.subplot2grid((1, num_columns), (0, 0), colspan=num_columns//3)  # Tree takes up 1/3 of the figure\n",
    "        ax.grid(False)\n",
    "\n",
    "        fig.subplots_adjust(wspace=0, hspace=0)\n",
    "\n",
    "\n",
    "        if options.labels:\n",
    "            fsize = 12 - 0.1*roary.shape[1] + 2\n",
    "            if fsize < 7:\n",
    "                fsize = 7\n",
    "            with plt.rc_context({'font.size': fsize}):\n",
    "                Phylo.draw(t, axes=ax, \n",
    "                           show_confidence=False,\n",
    "                           label_func=lambda x: str(x)[:10],\n",
    "                           xticks=([],), yticks=([],),\n",
    "                           ylabel=('',), xlabel=('',),\n",
    "                           xlim=(-mdist*0.1,mdist+mdist*0.45-mdist*roary.shape[1]*0.001),\n",
    "                           axis=('off',),\n",
    "                           title=('Tree\\n(%d strains)'%roary.shape[1],), \n",
    "                           do_show=False,\n",
    "                          )\n",
    "        else:\n",
    "            Phylo.draw(t, axes=ax, \n",
    "                       show_confidence=False,\n",
    "                       label_func=lambda x: None,\n",
    "                       xticks=([],), yticks=([],),\n",
    "                       ylabel=('',), xlabel=('',),\n",
    "                       xlim=(-mdist*0.1,mdist+mdist*0.1),\n",
    "                       axis=('off',),\n",
    "                       title=('Tree\\n(%d strains)'%roary.shape[1],), \n",
    "                       do_show=False,\n",
    "                      )\n",
    "\n",
    "\n",
    "    # Remove grid lines from each subplot \n",
    "for ax in axs_depth:\n",
    "    ax.grid(False)\n",
    "    ax.set_xticklabels([])\n",
    "    ax.set_yticklabels([])\n",
    "    ax.set_xticklabels([])\n",
    "    ax.set_yticklabels([])\n",
    "    ax.set_xlabel('')\n",
    "    ax.set_ylabel('')\n",
    "    ax.set_xticks([])\n",
    "    ax.set_yticks([])\n",
    "\n",
    "# Remove the space between the subplots\n",
    "plt.subplots_adjust(wspace=0, hspace=0)\n",
    "\n",
    "# Plot the coverage depth for each taxa\n",
    "for i, (ax, terminal) in enumerate(zip(axs_depth, t.get_terminals())):\n",
    "\n",
    "    # Get the name of the taxa\n",
    "    taxa_name = terminal.name\n",
    "\n",
    "    # Load the amplicon_positions dictionary\n",
    "    with open('amplicon_positions.pkl', 'rb') as f:\n",
    "        amplicon_positions = pickle.load(f)\n",
    "\n",
    "    # Check if the taxa has any amplicons\n",
    "    if taxa_name in amplicon_positions:\n",
    "        # Iterate over the amplicons\n",
    "        for p1loc, p2loc in amplicon_positions[taxa_name]:\n",
    "            # Create the coverage depth plot\n",
    "            ax.fill_between([p1loc, p2loc], [1, 1], color='#7c898b', alpha=1.0)\n",
    "\n",
    "    # Get the maximum genome length across all taxa\n",
    "    max_genome_length = max(genome_lengths.values())\n",
    "\n",
    "    # Set the x-axis limits to match the range of genome lengths\n",
    "    ax.set_xlim(0, max_genome_length)\n",
    "\n",
    "    # Get the coverage percentage for the taxa, default to 0 if not found\n",
    "    coverage_percentage = total_genome_coverages.get(taxa_name, 0)\n",
    "\n",
    "    # Get the position of the original axes\n",
    "    pos = ax.get_position()\n",
    "\n",
    "    # Create a new axes for the bar plot just to the right of the original axes\n",
    "    ax_bar = plt.axes([pos.x1 + 0.01, pos.y0, 0.05, pos.height])\n",
    "\n",
    "    # Plot the horizontal bar on the new axes\n",
    "    ax_bar.barh(0, coverage_percentage, height=2.0, color='#2d232e')\n",
    "\n",
    "    # Hide the y-axis\n",
    "    ax_bar.yaxis.set_visible(False)\n",
    "\n",
    "    # Set the x-axis limits to match the range of coverage percentages\n",
    "    ax_bar.set_xlim(0, 100)\n",
    "\n",
    "    # Set the y-axis limits to match the original axes\n",
    "    ax_bar.set_ylim(ax.get_ylim())\n",
    "\n",
    "    # Set x-label\n",
    "    ax_bar.set_xlabel('Coverage (%)')\n",
    "\n",
    "    # Remove axes spines \n",
    "    for s in ['top', 'right', 'left']:\n",
    "        ax_bar.spines[s].set_visible(False)\n",
    "\n",
    "    # Hide x-axis labels and ticks for all but the last ax_bar\n",
    "    if i < len(axs_depth) - 1:\n",
    "        ax_bar.xaxis.set_visible(False)\n",
    "        ax_bar.spines['bottom'].set_visible(False)\n",
    "\n",
    "    # Remove the title\n",
    "    ax.set_title('')\n",
    "\n",
    "    # Set the y-axis limits\n",
    "    ax.set_ylim(0, 1)  # Set y limit to 1 as the coverage is either 0 or 1\n",
    "\n",
    "    # Remove the x and y axis labels\n",
    "    ax.set_xticklabels([])\n",
    "    ax.set_yticklabels([])\n",
    "    ax.set_xticklabels([])\n",
    "    ax.set_yticklabels([])\n",
    "    ax.set_xlabel('')\n",
    "    ax.set_ylabel('')\n",
    "    ax.set_xticks([])\n",
    "    ax.set_yticks([])\n",
    "\n",
    "    # Remove grid \n",
    "    ax.grid(False)\n",
    "\n",
    "    # Remove seaborn grid\n",
    "    sns.despine(ax=ax, left=True, bottom=True)\n",
    "\n",
    "# Get the position of the last ax\n",
    "pos = ax.get_position()\n",
    "\n",
    "# Create a new axes for the genome coordinate line just below the last ax\n",
    "ax_genome = plt.axes([pos.x0, pos.y0, pos.width, 0.0002])\n",
    "\n",
    "# Hide the y-axis\n",
    "ax_genome.yaxis.set_visible(False)\n",
    "\n",
    "# Set the x-axis limits to match the range of genome lengths\n",
    "ax_genome.set_xlim(0, max_genome_length)\n",
    "\n",
    "# Set x-axis label\n",
    "ax_genome.set_xlabel('Genome position (mbp)')\n",
    "\n",
    "# Remove axes spines \n",
    "for s in ['top', 'right', 'left']:\n",
    "    ax_genome.spines[s].set_visible(False)\n",
    "            \n",
    "# Save the plot\n",
    "plt.savefig('pangenome_matrix_SP_reps_x.%s'%options.format, dpi=300)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
