{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc50ee06-71bd-4239-bec0-4bffa3581bef",
   "metadata": {},
   "outputs": [],
   "source": [
    "            ------------------------------------------------------ separate fwd and rev amplicon predictions --------------------------------------------------------------"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "89b99e8b-06fe-4afa-9bef-a7e1a30d8605",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "import yaml\n",
    "import numpy as np\n",
    "\n",
    "# load in config with GPSC paths\n",
    "with open(\"config/TB_reps.yaml\", 'r') as file:\n",
    "    config=yaml.safe_load(file)\n",
    "\n",
    "# extract GPSCs\n",
    "gpscs = config['samples'] \n",
    "\n",
    "# set params\n",
    "amplicon_stats = list()\n",
    "xlen = 2200\n",
    "\n",
    "total_genome_coverages={}\n",
    "amplicon_positions={}\n",
    "\n",
    "with open('output.txt', 'w') as f:\n",
    "\n",
    "    for gpsc, fasta_file in gpscs.items():\n",
    "        records = list(SeqIO.parse(fasta_file, \"fasta\"))\n",
    "\n",
    "        # calculate length of the genome for each GPSC\n",
    "        genome_length=sum(len(record.seq) for record in records)\n",
    "\n",
    "        # initialize amplicon genome coverage for each GPSC separately \n",
    "        total_genome_coverage=0\n",
    "\n",
    "        # initialize set of covered positions across each GPSC \n",
    "        covered_positions = set()\n",
    "\n",
    "        # intitialize list of amplicon positions \n",
    "        amplicon_positions[gpsc] = []\n",
    "\n",
    "        print(f\"Processing {gpsc}...\", file=f)  \n",
    "\n",
    "        # load the samtools depth file\n",
    "        fwd_depth_file = os.path.join(\"samtools_depth_indiv_primers\", f\"{gpsc}_fwd.depth\")\n",
    "        rev_depth_file = os.path.join(\"samtools_depth_indiv_primers\", f\"{gpsc}_rev.depth\")\n",
    "        \n",
    "        # load each depth file into a df\n",
    "        fwd_df = pd.read_csv(fwd_depth_file, sep=\"\\t\", names=[\"Ref\", \"Pos\", \"Depth\"])\n",
    "        fwd_df['Primer'] = 'fwd'\n",
    "\n",
    "        rev_df = pd.read_csv(rev_depth_file, sep=\"\\t\", names=[\"Ref\", \"Pos\", \"Depth\"])\n",
    "        rev_df['Primer'] = 'rev'\n",
    "\n",
    "        # combine the dfs\n",
    "        df = pd.concat([fwd_df, rev_df])\n",
    "\n",
    "        # filter for fwd and rev positions separately \n",
    "        fwd_primer_binding_sites = df[(df[\"Depth\"] == 1) & (df[\"Primer\"] == 'fwd')][\"Pos\"].tolist()\n",
    "\n",
    "        rev_primer_binding_sites = df[(df[\"Depth\"] == 1) & (df[\"Primer\"] == 'rev')][\"Pos\"].tolist()\n",
    "\n",
    "        def calculate_amplicons(primer_binding_sites, gpsc, xlen, amplicon_positions, amplicon_stats, covered_positions, total_genome_coverage, f, primer_direction):\n",
    "            for p1loc in primer_binding_sites:\n",
    "        # find the next primer binding site within xlen bases\n",
    "                p2loc = next((pos for pos in primer_binding_sites if p1loc < pos <= p1loc + xlen), None)\n",
    "\n",
    "                if p2loc is not None:\n",
    "            # amp stats\n",
    "                    amplicon_stats.append((gpsc, gpsc, gpsc, gpsc, gpsc, gpsc, 0, 0, 0, p1loc, p2loc, primer_direction))\n",
    "                    covered_positions.update(range(p1loc, p2loc+1))\n",
    "            # ID amp positions\n",
    "                    amplicon_positions[gpsc].append((p1loc, p2loc))\n",
    "                    print(f\"Detected amplicon from {p1loc} to {p2loc}.\", file=f)  \n",
    "\n",
    "            # get total genome coverage\n",
    "                    total_genome_coverage += p2loc - p1loc\n",
    "\n",
    "        # calculate predicted % coverage \n",
    "            coverage_percentage = (len(covered_positions) / genome_length) * 100\n",
    "\n",
    "        # update dictionary \n",
    "            total_genome_coverages[gpsc] = coverage_percentage\n",
    "\n",
    "        # summarize fwd amplicons\n",
    "        calculate_amplicons(fwd_primer_binding_sites, gpsc, xlen, amplicon_positions, amplicon_stats, covered_positions, total_genome_coverage, f, 'fwd')\n",
    "        # summarize rev amplicons\n",
    "        calculate_amplicons(rev_primer_binding_sites, gpsc, xlen, amplicon_positions, amplicon_stats, covered_positions, total_genome_coverage, f, 'rev')\n",
    "\n",
    "colnames = [\"pid1\", \"pid2\", \"set1\", \"set2\",\n",
    "            \"pseq1\", \"pseq2\",\n",
    "            \"max_hdist\", \"hdist1\", \"hdist2\",\n",
    "            \"p1loc\", \"p2loc\", \"PrimerDirection\"]\n",
    "coltypes = [\"<U30\", \"<U30\", \"<U30\", \"<U30\",\n",
    "            \"<U30\", \"<U30\",\n",
    "            float, float, float,\n",
    "            int, int, \"<U3\"]\n",
    "\n",
    "dt = {'names': colnames, 'formats': coltypes}\n",
    "\n",
    "amplicon_statsnp = np.array(amplicon_stats,\n",
    "                     dtype=dt)\n",
    "\n",
    "np.save(\"amplicon_statstab.npy\", amplicon_statsnp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "12e8e49b-c9d3-486c-ac46-71f99dc8bd48",
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv\n",
    "\n",
    "# Initialize a dictionary to group amplicons by sequence and primer type\n",
    "grouped_amplicons = {}\n",
    "\n",
    "# Define the maximum allowed size for amplicons\n",
    "max_amplicon_size = 2000\n",
    "\n",
    "# Iterate over amplicon_stats, assuming the last element is primer_direction, and the first is gpsc\n",
    "for amplicon in amplicon_stats:\n",
    "    *_, p1loc, p2loc, primer_direction = amplicon\n",
    "    gpsc = amplicon[0]\n",
    "    key = (gpsc, primer_direction)\n",
    "    if key not in grouped_amplicons:\n",
    "        grouped_amplicons[key] = []\n",
    "    grouped_amplicons[key].append((p1loc, p2loc))\n",
    "\n",
    "# Initialize list for contiguous amplicons\n",
    "contiguous_amplicons = []\n",
    "\n",
    "# Process each group for contiguity\n",
    "for (gpsc, primer_direction), positions in grouped_amplicons.items():\n",
    "    positions.sort()\n",
    "    start, end = positions[0]\n",
    "    \n",
    "    for current_start, current_end in positions[1:]:\n",
    "        # Check if contiguous and does not exceed max amplicon size\n",
    "        if current_start <= end + 1 and (current_end - start) <= max_amplicon_size:\n",
    "            end = max(end, current_end)\n",
    "        else:\n",
    "            contiguous_amplicons.append([gpsc, start, end, primer_direction])\n",
    "            start, end = current_start, current_end\n",
    "    # After processing all positions, append the last contiguous segment\n",
    "    contiguous_amplicons.append([gpsc, start, end, primer_direction])\n",
    "\n",
    "# Write contiguous amplicons to CSV\n",
    "with open('amplicon_positions.csv', 'w', newline='') as csvfile:\n",
    "    csvwriter = csv.writer(csvfile)\n",
    "    csvwriter.writerow([\"Sequence\", \"Start\", \"End\", \"Primer Type\"])\n",
    "    for amplicon in contiguous_amplicons:\n",
    "        csvwriter.writerow(amplicon)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "79f1ddc9-c924-4d74-bb5f-7745bf8f7297",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# create csv file containing predicted genome coverage for each sequence based on predicted amplicon coverage\n",
    "total_genome_coverages_df = pd.DataFrame([total_genome_coverages])\n",
    "\n",
    "total_genome_coverages_df.to_csv('genome_coverage_pc.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "0d221bda",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ERR144577: 94.26547606957885%\n",
      "ERR212146: 94.25506116650568%\n",
      "ERR181435: 94.3077754119962%\n",
      "SRR998654: 94.2374807773216%\n",
      "ERR161044: 94.26067335448154%\n",
      "ERR234202: 94.25072140831597%\n",
      "SRR1019139: 94.25973079801055%\n",
      "SRR1163294: 94.2444317807285%\n",
      "SRR1180462: 94.23058657973675%\n",
      "SRR1172044: 94.2951933896527%\n",
      "SRR671749: 94.29532559441637%\n",
      "SRR1173640: 94.26387550860808%\n",
      "H37Rv: 94.29513375398841%\n",
      "NC_019950: 89.43826698968014%\n"
     ]
    }
   ],
   "source": [
    "for gpsc, coverage in total_genome_coverages.items():\n",
    "    print(f\"{gpsc}: {coverage}%\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "8b6f337d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ERR144577: 4411456 bp\n",
      "ERR212146: 4411483 bp\n",
      "ERR181435: 4411509 bp\n",
      "SRR998654: 4411456 bp\n",
      "ERR161044: 4411406 bp\n",
      "ERR234202: 4411510 bp\n",
      "SRR1019139: 4411483 bp\n",
      "SRR1163294: 4411554 bp\n",
      "SRR1180462: 4411523 bp\n",
      "SRR1172044: 4411508 bp\n",
      "SRR671749: 4411470 bp\n",
      "SRR1173640: 4411550 bp\n",
      "H37Rv: 4411532 bp\n",
      "NC_019950: 4432426 bp\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "# Load the config file\n",
    "with open(\"config/TB_reps.yaml\", 'r') as file:\n",
    "    config=yaml.safe_load(file)\n",
    "\n",
    "# Extract the GPSCs\n",
    "gpscs = config['samples']\n",
    "\n",
    "# Initialize a dictionary to store the genome lengths\n",
    "genome_lengths = {}\n",
    "\n",
    "# Calculate and store the length of each sequence\n",
    "for gpsc, fasta_file in gpscs.items():\n",
    "    records = list(SeqIO.parse(fasta_file, \"fasta\"))\n",
    "    genome_length = sum(len(record.seq) for record in records)\n",
    "    genome_lengths[gpsc] = genome_length  # Store the genome length in the dictionary\n",
    "\n",
    "#Print the genome lengths\n",
    "for gpsc, length in genome_lengths.items():\n",
    "    print(f\"{gpsc}: {length} bp\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "7bb8b9e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv\n",
    "import yaml\n",
    "from Bio import SeqIO\n",
    "\n",
    "# Load the config file\n",
    "with open(\"config/TB_reps.yaml\", 'r') as file:\n",
    "    config = yaml.safe_load(file)\n",
    "\n",
    "# Extract the GPSCs\n",
    "gpscs = config['samples']\n",
    "\n",
    "# Open a CSV file to write\n",
    "with open('assembly_lengths.csv', mode='w', newline='') as csv_file:\n",
    "    # Create a CSV writer object\n",
    "    csv_writer = csv.writer(csv_file)\n",
    "    \n",
    "    # Calculate and store the length of each sequence, then write to CSV\n",
    "    for gpsc, fasta_file in gpscs.items():\n",
    "        records = list(SeqIO.parse(fasta_file, \"fasta\"))\n",
    "        genome_length = sum(len(record.seq) for record in records)\n",
    "        # Write the ID, start (0), and the length to the CSV file\n",
    "        csv_writer.writerow([gpsc, 0, genome_length])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
